README
NAME

    WebService::Stripe - Stripe API bindings

VERSION

    version 0.0600

SYNOPSIS

        my $stripe = WebService::Stripe->new(
            api_key => 'secret',
            version => '2014-11-05', # optional
        );
        my $customer = $stripe->get_customer('cus_57eDUiS93cycyH');

METHODS

 get_customer

        get_customer($id)

    Returns the customer for the given id.

 create_customer

        create_customer($data)

    Creates a customer. The $data hashref is optional. Returns the
    customer.

    Example:

        $customer = $stripe->create_customer({ email => 'bob@foo.com' });

 update_customer

        update_customer($id, $data)

    Updates a customer. Returns the updated customer.

    Example:

        $customer = $stripe->update_customer($id, { description => 'foo' });

 get_customers

        get_customers(query => $query)

    Returns a list of customers. The query param is optional.

 next

        next($collection)

    Returns the next page of results for the given collection.

    Example:

        my $customers = $stripe->get_customers;
        ...
        while ($customers = $stripe->next($customers)) {
            ...
        }

 create_card

        create_card($data, customer_id => 'cus_123')

 get_charge

        get_charge($id)

    Returns the charge for the given id.

 create_charge

        create_charge($data)

    Creates a charge.

 capture_charge

        capture_charge($id, data => $data)

    Captures the charge with the given id. The data param is optional.

 refund_charge

        refund_charge($id, data => $data)

    Refunds the charge with the given id. The data param is optional.

 get_token

        get_token($id)

 create_token

        create_token($data)

 get_account

        get_account($id)

 create_account

        create_account($data)

 update_account

        update_account($id, data => $data)

 add_bank

        add_bank($data, account_id => $account_id)

    Add a bank to an account.

    Example:

        my $account = $stripe->create_account({
            managed => 'true',
            country => 'CA',
        });
    
        my $bank = $stripe->add_bank(
            {
                'bank_account[country]'        => 'CA',
                'bank_account[currency]'       => 'cad',
                'bank_account[routing_number]' => '00022-001',
                'bank_account[account_number]' => '000123456789',
            },
            account_id => $account->{id},
        );
    
        # or add a tokenised bank
    
        my $bank_token = $stripe->create_token({
            'bank_account[country]'        => 'CA',
            'bank_account[currency]'       => 'cad',
            'bank_account[routing_number]' => '00022-001',
            'bank_account[account_number]' => '000123456789',
        });
    
        $stripe->add_bank(
            { bank_account => $bank_token->{id} },
            account_id => $account->{id},
        );

 update_bank

        update_bank($id, account_id => $account_id, data => $data)

 create_transfer

        create_transfer($data)

 get_transfer

        get_transfer($id)

 get_transfers

        get_transfers(query => $query)

 update_transfer

        update_transfer($id, data => $data)

 cancel_transfer

        cancel_transfer($id)

AUTHOR

    Naveed Massjouni <naveed@vt.edu>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Tilt, Inc.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



CHANGES
==================================================
Changes from 2004-12-07 00:00:00 +0000 to present.
==================================================

-------------------------------------------
version 0.0600 at 2014-12-05 10:10:22 +0000
-------------------------------------------

  Change: defa147f1f71fd38653d7efc3bd6477e1f1c0379
  Author: Naveed Massjouni <naveed@vt.edu>
  Date : 2014-12-05 05:08:02 +0000

    v0.0600 added support for transfers 

-------------------------------------------
version 0.0500 at 2014-12-04 08:40:31 +0000
-------------------------------------------

  Change: da82027dfbbd44b7c5cb205e1fcf15096fc3b41c
  Author: Naveed Massjouni <naveed@vt.edu>
  Date : 2014-12-04 03:40:31 +0000

    added more bank methods 

-------------------------------------------
version 0.0400 at 2014-12-01 10:52:32 +0000
-------------------------------------------

  Change: 5beeefac1cd62b8de7220a00ba0346342099eaf5
  Author: Naveed Massjouni <naveed@vt.edu>
  Date : 2014-12-01 05:52:32 +0000

    v0.0400 added support for the new marketplace api accounts and banks

    Added get_token, create_account, get_account, update_account, and
    add_bank. 

  Change: b723faa726f885c74518d03d458e5e81f65e4b4e
  Author: Naveed Massjouni <naveed@vt.edu>
  Date : 2014-11-29 08:26:02 +0000

    v0.0300 added the refund_charge method 

-------------------------------------------
version 0.0200 at 2014-11-21 08:50:48 +0000
-------------------------------------------

  Change: a43ecabdccfb11ef6456ef6df063c44b342f8803
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-21 03:50:48 +0000

    v0.0200 added get_charge and capture_charge 

  Change: f447aa1533acde96ce1c0cac6746b1cb1c1d86ef
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-19 02:39:16 +0000

    v0.0100 

  Change: d028e7f0f6936cbfa5206823366e38c0fe43babe
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-19 01:23:11 +0000

    create_card 

  Change: 84b05c98dc988ec5bca3587e4eb3966d048e1540
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-18 10:35:37 +0000

    stripe_account header 

  Change: 1bbc1123cff98ce96786520a6b601a593a6a78aa
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-17 04:09:51 +0000

    added create_charge and create_token 

  Change: 94e7c217b79fc2a0a5ed4d74a73f64cf1da5a660
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-13 23:37:22 +0000

    next 

  Change: d3a3df93b931a840100e269fdb29319b48f2bc6a
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-12 17:12:44 +0000

    require newer WebService::Client 

  Change: 951e23fdd2f768757dc39937a0da0824cc8676e5
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-12 16:00:10 +0000

    added create_customer and update_customer and tests 

  Change: 2507c608bfd38e8411d513d70130ecd13ec73bf9
  Author: Naveed Massjouni <naveedm9@gmail.com>
  Date : 2014-11-11 17:57:45 +0000

    initial commit 

================
End of releases.
================


lib/WebService/Stripe.pm
package WebService::Stripe;
use Moo;
with 'WebService::Client';

our $VERSION = '0.0600'; # VERSION

use Carp qw(croak);
use Method::Signatures;

has api_key => (
    is       => 'ro',
    required => 1,
);

has version => (
    is      => 'ro',
    default => '2014-11-05',
);

has '+base_url' => ( default => 'https://api.stripe.com' );

has '+content_type' => ( default => 'application/x-www-form-urlencoded' );

method BUILD(@args) {
    $self->ua->default_headers->authorization_basic( $self->api_key, '' );
    $self->ua->default_header( 'Stripe-Version' => '2014-11-05' );
}

method next(HashRef $thing, HashRef :$query) {
    $query ||= {};
    return undef unless $thing->{has_more};
    my $starting_after = $thing->{data}[-1]{id} or return undef;
    return $self->get( $thing->{url},
        { %$query, starting_after => $starting_after } );
}

method create_customer(HashRef $data={}) {
    return $self->post( "/v1/customers", $data );
}

method get_customer(Str $id) {
    return $self->get( "/v1/customers/$id" );
}

method update_customer(Str $id, HashRef $data) {
    return $self->post( "/v1/customers/$id", $data );
}

method get_customers(HashRef :$query) {
    return $self->get( "/v1/customers", $query );
}

method create_card(HashRef $data, :$customer_id!) {
    return $self->post( "/v1/customers/$customer_id/cards", $data );
}

method get_charge(Str $id) {
    return $self->get( "/v1/charges/$id" );
}

method create_charge(HashRef $data) {
    return $self->post( "/v1/charges", $data );
}

method capture_charge(Str $id, HashRef :$data) {
    return $self->post( "/v1/charges/$id/capture", $data );
}

method refund_charge(Str $id, HashRef :$data) {
    return $self->post( "/v1/charges/$id/refunds", $data );
}

method create_token(HashRef $data) {
    return $self->post( "/v1/tokens", $data );
}

method get_token(Str $id) {
    return $self->get( "/v1/tokens/$id" );
}

method create_account(HashRef $data) {
    return $self->post( "/v1/accounts", $data );
}

method get_account(Str $id) {
    return $self->get( "/v1/accounts/$id" );
}

method update_account(Str $id, HashRef :$data!) {
    return $self->post( "/v1/accounts/$id", $data );
}

method add_bank(HashRef $data, Str :$account_id!) {
    return $self->post( "/v1/accounts/$account_id/bank_accounts", $data );
}

method update_bank(Str $id, Str :$account_id!, HashRef :$data!) {
    return $self->post( "/v1/accounts/$account_id/bank_accounts/$id", $data );
}

method delete_bank(Str $id, Str :$account_id!) {
    return $self->delete( "/v1/accounts/$account_id/bank_accounts/$id" );
}

method get_banks(Str :$account_id!) {
    return $self->get( "/v1/accounts/$account_id/bank_accounts" );
}

method create_transfer(HashRef $data) {
    return $self->post( "/v1/transfers", $data );
}

method get_transfer(Str $id) {
    return $self->get( "/v1/transfers/$id" );
}

method get_transfers(HashRef :$query) {
    return $self->get( "/v1/transfers", $query );
}

method update_transfer(Str $id, HashRef :$data!) {
    return $self->post( "/v1/transfers/$id", $data );
}

method cancel_transfer(Str $id) {
    return $self->post( "/v1/transfers/$id/cancel" );
}

# ABSTRACT: Stripe API bindings


1;

__END__

=pod

=encoding UTF-8

=head1 NAME

WebService::Stripe - Stripe API bindings

=head1 VERSION

version 0.0600

=head1 SYNOPSIS

    my $stripe = WebService::Stripe->new(
        api_key => 'secret',
        version => '2014-11-05', # optional
    );
    my $customer = $stripe->get_customer('cus_57eDUiS93cycyH');

=head1 METHODS

=head2 get_customer

    get_customer($id)

Returns the customer for the given id.

=head2 create_customer

    create_customer($data)

Creates a customer.
The C<$data> hashref is optional.
Returns the customer.

Example:

    $customer = $stripe->create_customer({ email => 'bob@foo.com' });

=head2 update_customer

    update_customer($id, $data)

Updates a customer.
Returns the updated customer.

Example:

    $customer = $stripe->update_customer($id, { description => 'foo' });

=head2 get_customers

    get_customers(query => $query)

Returns a list of customers.
The query param is optional.

=head2 next

    next($collection)

Returns the next page of results for the given collection.

Example:

    my $customers = $stripe->get_customers;
    ...
    while ($customers = $stripe->next($customers)) {
        ...
    }

=head2 create_card

    create_card($data, customer_id => 'cus_123')

=head2 get_charge

    get_charge($id)

Returns the charge for the given id.

=head2 create_charge

    create_charge($data)

Creates a charge.

=head2 capture_charge

    capture_charge($id, data => $data)

Captures the charge with the given id.
The data param is optional.

=head2 refund_charge

    refund_charge($id, data => $data)

Refunds the charge with the given id.
The data param is optional.

=head2 get_token

    get_token($id)

=head2 create_token

    create_token($data)

=head2 get_account

    get_account($id)

=head2 create_account

    create_account($data)

=head2 update_account

    update_account($id, data => $data)

=head2 add_bank

    add_bank($data, account_id => $account_id)

Add a bank to an account.

Example:

    my $account = $stripe->create_account({
        managed => 'true',
        country => 'CA',
    });

    my $bank = $stripe->add_bank(
        {
            'bank_account[country]'        => 'CA',
            'bank_account[currency]'       => 'cad',
            'bank_account[routing_number]' => '00022-001',
            'bank_account[account_number]' => '000123456789',
        },
        account_id => $account->{id},
    );

    # or add a tokenised bank

    my $bank_token = $stripe->create_token({
        'bank_account[country]'        => 'CA',
        'bank_account[currency]'       => 'cad',
        'bank_account[routing_number]' => '00022-001',
        'bank_account[account_number]' => '000123456789',
    });

    $stripe->add_bank(
        { bank_account => $bank_token->{id} },
        account_id => $account->{id},
    );

=head2 update_bank

    update_bank($id, account_id => $account_id, data => $data)

=head2 create_transfer

    create_transfer($data)

=head2 get_transfer

    get_transfer($id)

=head2 get_transfers

    get_transfers(query => $query)

=head2 update_transfer

    update_transfer($id, data => $data)

=head2 cancel_transfer

    cancel_transfer($id)

=head1 AUTHOR

Naveed Massjouni <naveed@vt.edu>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Tilt, Inc.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
